function crashes when environmentconfig reference is not found

<!--
Thank you for helping to improve Crossplane!

Please be sure to search for open issues before raising a new one. We use issues
for bug reports and feature requests. Please find us at https://slack.crossplane.io
for questions, support, and discussion.
-->

### What happened?
Function crashes when referencing to non-existing environment config.


### How can we reproduce it?
Test environment config:
```
apiVersion: apiextensions.crossplane.io/v1alpha1
metadata:
  name: environment
data:
  branch: feature/dev
  environment: dev
kind: EnvironmentConfig
```

Setup step in composition that reads something similar to:
```
  - step: environment-configs
    functionRef:
      name: function-environment-configs
    input:
      apiVersion: environmentconfigs.fn.crossplane.io/v1beta1
      kind: Input
      spec:
        environmentConfigs:
          - type: Reference
            ref:
              name: does_not_exist
```

Submit and you'll get:

```
2024-03-28T22:35:58.242Z        INFO    fn/fn.go:40     Running function        {"tag": ""}
2024-03-28T22:35:58.242Z        DEBUG   fn/fn.go:72     No extra resources specified, exiting   {"requirements": "extra_resources:{key:\"environment-config-0\" value:{api_version:\"apiextensions.crossplane.io/v1alpha1\" kind:\"EnvironmentConfig\" match_name:\"does_not_exist\"}}"}
2024-03-28T22:35:58.246Z        INFO    fn/fn.go:40     Running function        {"tag": ""}
2024-03-28T22:35:58.246Z        DEBUG   fn/fn.go:83     Loaded Composition environment from Function context    {"context-key": "apiextensions.crossplane.io/environment"}
panic: runtime error: index out of range [0] with length 0

goroutine 5363 [running]:
main.getSelectedEnvConfigs(0xc0004c2500, 0x175a8a3?)
        /fn/fn.go:150 +0x7ab
main.(*Function).RunFunction(0xc00041e840, {0xc0002a0660?, 0x499ae6?}, 0xc0002a0660)
        /fn/fn.go:92 +0x328
github.com/crossplane/function-sdk-go/proto/v1beta1._FunctionRunnerService_RunFunction_Handler({0x1558b20?, 0xc00041e840}, {0x196f8c0, 0xc00052c0c0}, 0xc0001e4c00, 0x0)
        /go/pkg/mod/github.com/crossplane/function-sdk-go@v0.2.0/proto/v1beta1/run_function_grpc.pb.go:104 +0x169
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000fa000, {0x196f8c0, 0xc00052c030}, {0x1975140, 0xc0005029c0}, 0xc000417b00, 0xc00058a630, 0x248b7d0, 0x0)
        /go/pkg/mod/google.golang.org/grpc@v1.61.0/server.go:1385 +0xe03
google.golang.org/grpc.(*Server).handleStream(0xc0000fa000, {0x1975140, 0xc0005029c0}, 0xc000417b00)
        /go/pkg/mod/google.golang.org/grpc@v1.61.0/server.go:1796 +0xfec
google.golang.org/grpc.(*Server).serveStreams.func2.1()
        /go/pkg/mod/google.golang.org/grpc@v1.61.0/server.go:1029 +0x8b
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 8
        /go/pkg/mod/google.golang.org/grpc@v1.61.0/server.go:1040 +0x135
```


### What environment did it happen in?
Function version: 0.0.4 - 0.0.6
Crossplane: 1.15.0
(testing locally via crank/crossplane validation and in k8s cluster)

--- a/fn.go
+++ b/fn.go
@@ -141,8 +141,11 @@ func getSelectedEnvConfigs(in *v1beta1.Input, extraResources map[string][]resour
 		}
 		switch config.GetType() {
 		case v1beta1.EnvironmentSourceTypeReference:
-			if len(resources) == 0 && in.Spec.Policy.IsResolutionPolicyOptional() {
-				continue
+			if len(resources) == 0 {
+				if in.Spec.Policy.IsResolutionPolicyOptional() {
+					continue
+				}
+				return nil, errors.Errorf("Required environment config %q not found", extraResName)
 			}
 			if len(resources) > 1 {
 				return nil, errors.Errorf("expected exactly one extra resource %q, got %d", extraResName, len(resources))

--- a/fn_test.go
+++ b/fn_test.go
@@ -6,7 +6,7 @@ import (
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/google/go-cmp/cmp/cmpopts"
-	"google.golang.org/protobuf/testing/protocmp"
+	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/types/known/durationpb"
 	"google.golang.org/protobuf/types/known/structpb"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
@@ -401,6 +401,65 @@ func TestRunFunction(t *testing.T) {
 				},
 			},
 		},
+		"RequestEnvironmentConfigsNotFoundRequired": {
+			reason: "The Function should return fatal if a required EnvironmentConfig is not found",
+			args: args{
+				req: &fnv1beta1.RunFunctionRequest{
+					Meta: &fnv1beta1.RequestMeta{Tag: "hello"},
+					Observed: &fnv1beta1.State{
+						Composite: &fnv1beta1.Resource{
+							Resource: resource.MustStructJSON(`{
+								"apiVersion": "test.crossplane.io/v1alpha1",
+								"kind": "XR",
+								"metadata": {
+									"name": "my-xr"
+								}
+							}`),
+						},
+					},
+					ExtraResources: map[string]*fnv1beta1.Resources{
+						"environment-config-0": {
+							Items: []*fnv1beta1.Resource{},
+						},
+					},
+					Input: resource.MustStructJSON(`{
+						"apiVersion": "template.fn.crossplane.io/v1beta1",
+						"kind": "Input",
+						"spec": {
+							"environmentConfigs": [
+								{	
+									"type": "Reference",
+									"ref": {	
+										"name": "my-env-config"
+									}
+								}
+							]
+						}
+					}`),
+				},
+			},
+			want: want{
+				rsp: &fnv1beta1.RunFunctionResponse{
+					Meta: &fnv1beta1.ResponseMeta{Tag: "hello", Ttl: durationpb.New(response.DefaultTTL)},
+					Results: []*fnv1beta1.Result{
+						{
+							Severity: fnv1beta1.Severity_SEVERITY_FATAL,
+						},
+					},
+					Requirements: &fnv1beta1.Requirements{
+						ExtraResources: map[string]*fnv1beta1.ResourceSelector{
+							"environment-config-0": {
+								ApiVersion: "apiextensions.crossplane.io/v1alpha1",
+								Kind:       "EnvironmentConfig",
+								Match: &fnv1beta1.ResourceSelector_MatchName{
+									MatchName: "my-env-config",
+								},
+							},
+						},
+					},
+				},
+			},
+		},
 		"MergeEnvironmentConfigs": {
 			reason: "The Function should merge the provided EnvironmentConfigs",
 			args: args{
@@ -529,7 +588,22 @@ func TestRunFunction(t *testing.T) {
 			f := &Function{log: logging.NewNopLogger()}
 			rsp, err := f.RunFunction(tc.args.ctx, tc.args.req)
 
-			if diff := cmp.Diff(tc.want.rsp, rsp, protocmp.Transform()); diff != "" {
+			diff := cmp.Diff(tc.want.rsp, rsp, cmpopts.AcyclicTransformer("toJsonWithoutResultMessages", func(r *fnv1beta1.RunFunctionResponse) []byte {
+				// We don't care about messages.
+				// cmptopts.IgnoreField wasn't working with protocmp.Transform
+				// We can't split this to another transformer as
+				// transformers are applied not in order but as soon as they
+				// match the type, which are walked from the root (RunFunctionResponse).
+				for _, result := range r.Results {
+					result.Message = ""
+				}
+				out, err := protojson.Marshal(r)
+				if err != nil {
+					t.Fatalf("cannot marshal %T to JSON: %s", r, err)
+				}
+				return out
+			}))
+			if diff != "" {
 				t.Errorf("%s\nf.RunFunction(...): -want rsp, +got rsp:\n%s", tc.reason, diff)
 			}
 

